[JSC] Further optimize JSRopeString::resolveRope

Constellation opened the pull request:
#### 4d816460b765acd8aef90ab474615850b91ecc35
<pre>
[JSC] Further optimize JSRopeString::resolveRope
<a href="https://bugs.webkit.org/show_bug.cgi?id=259203">https://bugs.webkit.org/show_bug.cgi?id=259203</a>
rdar://112241674

Reviewed by Mark Lam.

This patch furhter improves JSRopeString::resolveRope and jsAtomString.

1. We crafted JSRopeString::resolveToBuffer very carefully so that it becomes tail calls well.
   This allows clang to make it loop instead of calls for recursive resolveToBuffer calls, which
   is significantly efficient. We ensure this condition is met by using MUST_TAIL_CALL C++ attribute,
   which is available in clang.
2. We tweak jsAtomString to make it better in terms of performance. It turned out these duplicate part
   is mandatory to make it super efficient, which is observed by JetStream2/WSL.

Simple rope resolution gets 3-4% improvement.

                                   ToT                     Patched

rope-resolve-recursive       21.3292+-0.6975     ^     20.5524+-0.0687        ^ definitely 1.0378x faster

* Source/JavaScriptCore/runtime/JSString.cpp:
(JSC::JSRopeString::resolveRopeInternalNoSubstring const):
* Source/JavaScriptCore/runtime/JSString.h:
* Source/JavaScriptCore/runtime/JSStringInlines.h:
(JSC::JSRopeString::resolveToBufferSlow):
(JSC::JSRopeString::resolveToBuffer):
(JSC::jsAtomString):
(JSC::JSStringFibers::JSStringFibers): Deleted.
(JSC::JSStringFibers::fiber const): Deleted.
(JSC::JSStringFibers::fiber0 const): Deleted.
(JSC::JSStringFibers::fiber1 const): Deleted.
(JSC::JSStringFibers::fiber2 const): Deleted.
(): Deleted.
* Source/WTF/wtf/Compiler.h:

Canonical link: <a href="https://commits.webkit.org/266060@main">https://commits.webkit.org/266060@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/39683a0a0264ae19f5e0c66fc798d2289cac6c16

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12724 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13048 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13376 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14463 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12168 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15552 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13069 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14871 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12888 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/13632 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10769 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14908 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10919 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11517 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/18595 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/10819 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/11994 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11685 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14881 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/12046 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/12146 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/10058 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/12779 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11401 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3355 "Passed tests") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/15721 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/13138 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1444 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/11995 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3143 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/runtime/JSStringInlines.h:
```diff
@@ -138,22 +122,24 @@ class JSStringFibers {
+    if (fiber1)
+        workQueue.append(fiber1);
+    if (fiber2)
+        workQueue.append(fiber2);
```
Per our offline discussion, this can be the following instead to avoid unnecessary checks:
```
    workQueue.append(fiber0);
    if (fiber1) {
        workQueue.append(fiber1);
        if (fiber2)
            workQueue.append(fiber2);
    }
```

Constellation replied:
Changed.

MenloDorian commented on Source/JavaScriptCore/runtime/JSStringInlines.h:
```diff
@@ -138,22 +122,24 @@ class JSStringFibers {
+    if (fiber2)
+        workQueue.append(fiber2);
 
     while (!workQueue.isEmpty()) {
```
Hmmm, the fact that this is using a while loop instead of a do-while loop makes me wonder if we every make a rope with 0 fibers.  Should this be a do-while instead?

Constellation replied:
Probably either is fine, but changed.

MenloDorian reviewed (approved):
r=me

Constellation committed:
[JSC] Further optimize JSRopeString::resolveRope
https://bugs.webkit.org/show_bug.cgi?id=259203
rdar://112241674

Reviewed by Mark Lam.

This patch furhter improves JSRopeString::resolveRope and jsAtomString.

1. We crafted JSRopeString::resolveToBuffer very carefully so that it becomes tail calls well.
   This allows clang to make it loop instead of calls for recursive resolveToBuffer calls, which
   is significantly efficient. We ensure this condition is met by using MUST_TAIL_CALL C++ attribute,
   which is available in clang.
2. We tweak jsAtomString to make it better in terms of performance. It turned out these duplicate part
   is mandatory to make it super efficient, which is observed by JetStream2/WSL.

Simple rope resolution gets 3-4% improvement.

                                   ToT                     Patched

rope-resolve-recursive       21.3292+-0.6975     ^     20.5524+-0.0687        ^ definitely 1.0378x faster

* Source/JavaScriptCore/runtime/JSString.cpp:
(JSC::JSRopeString::resolveRopeInternalNoSubstring const):
* Source/JavaScriptCore/runtime/JSString.h:
* Source/JavaScriptCore/runtime/JSStringInlines.h:
(JSC::JSRopeString::resolveToBufferSlow):
(JSC::JSRopeString::resolveToBuffer):
(JSC::jsAtomString):
(JSC::JSStringFibers::JSStringFibers): Deleted.
(JSC::JSStringFibers::fiber const): Deleted.
(JSC::JSStringFibers::fiber0 const): Deleted.
(JSC::JSStringFibers::fiber1 const): Deleted.
(JSC::JSStringFibers::fiber2 const): Deleted.
(): Deleted.
* Source/WTF/wtf/Compiler.h:

Canonical link: https://commits.webkit.org/266060@main

kraj commented on Source/WTF/wtf/Compiler.h:
```diff
@@ -281,6 +281,18 @@
+
+#if !defined(MUST_TAIL_CALL) && defined(__cplusplus) && defined(__has_cpp_attribute)
+#if __has_cpp_attribute(clang::musttail)
+#define MUST_TAIL_CALL [[clang::musttail]]
```
this does not work well with clang 17 on arm backend. see https://github.com/llvm/llvm-project/issues/67767#issuecomment-1741736944
